README.md

```diff
@@ -10,268 +10,475 @@ hard-to-find defects early in the application development lifecycle.
+        ```
+      * Certificate authentication:
+        ```
+        withCredentials([file(credentialsId: params.CSONAR_USER_CERT_ID, variable: 'CSONAR_USER_CERT'),
+                          file(credentialsId: params.CSONAR_USER_PRIVKEY_ID, variable: 'CSONAR_USER_PRIVKEY')]) { ... }
+        ```
+    1. Set up environment variable `CSONAR_CMD_ARGS` to collect together the various `codesonar` command line elements.
+       ```
+       env.CSONAR_CMD_ARGS = "$CSONAR_HUB_ADDRESS \
+                              $CSONAR_WORKDIR/$CSONAR_PROJECT_NAME \
+                              -project $CSONAR_PROJECT_NAME \
+                              -auth password -hubuser $CSONAR_HUBUSER -hubpwfile $CSONAR_HUBPWFILE \
+                              -foreground"
+       ```
+       * For certificate-based authentication, replace the `-auth` line with:
+         ```
+         -auth certificate -hubcert $CSONAR_USER_CERT -hubkey $CSONAR_USER_PRIVKEY \
+         ```
+       * For CodeSonar SaaS, replace the `-foreground` flag with:
+         ```
+         -remote /saas/*
+         ```
+       * For (non-SaaS) remote-managed analyses, replace the `-foreground` flag with:
+         ```
+         -remote <analysis-launchd>
+         ```
+         Where `<analysis-launchd>` specifies a suitable launch daemon or launchd group that is connected to your hub and can perform the analysis.
+       * If you want to specify any addition `codesonar` command line options, add them to this setting.
+         See [Optional Extras for $CSONAR_CMD_ARGS](#optional-extras-for-csonar_cmd_args), below, for some ideas.
+
+       **MANUAL:** Using CodeSonar > Command Line Build/Analysis
+    1. Edit the build steps to incorporate the CodeSonar build/analysis.
+
+       | Project Language | Editing the Build Steps |
+       |------------------|---------------------------|
+       | **C, C++**       | For every existing build step that involves C/C++ compilation, edit the build step to incorporate the CodeSonar build/analysis command. If the current build step or steps contain one command that involves C/C++ compilation, this will involve constructing a single `codesonar analyze` command. Otherwise there are two possible approaches: <ul><li>Accumulate components into a CodeSonar project by constructing a `codesonar build` command for each software build command that involves C/C++ compilation, then add a final `codesonar analyze` command to analyze the project.<br><i>or</i></li><li>Replace the text of the build step or steps with an invocation of a shell script or batch file with equivalent contents, then construct a single `codesonar analyze` command based on that invocation. </li></ul><p>See [Example 1](#example-1-cc-project-jenkins-build-steps-include-one-command-that-involves-cc-compilation) and [Example 2](#example-2-cc-project-jenkins-build-steps-include-multiple-commands-that-involve-cc-compilation).</p> |
```

Comment:
The hyperlinks for "Example 1" and "Example 2" don't seem to work.  You should be able to use an `<a name="target">` tag to define the link target name, which would make all of the internal links in this document more robust.  The `a` tag doesn't always work, but if you put it at the end of the heading line, it seems to be effective.